Reduce unsafeness in ImageDocument

annevk opened the pull request:
#### dfa351e2eabb50dadbe62445667a91dfe8d79c97
<pre>
Reduce unsafeness in ImageDocument
<a href="https://bugs.webkit.org/show_bug.cgi?id=304321">https://bugs.webkit.org/show_bug.cgi?id=304321</a>

Reviewed by Chris Dumez and Ryosuke Niwa.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a> and
adopt CachedResourceHandle while here.

Canonical link: <a href="https://commits.webkit.org/304649@main">https://commits.webkit.org/304649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64ff0b21ea08eb06bf263ccc33bd52b2ff7cc82b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47364 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143791 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9d58810b-f7a2-47e7-b0b6-fe1ddbe281b5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137954 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9113 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104030 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3493e8d4-6bf7-49b4-b04b-519dcdd9e81f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6621 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84879 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/05a67f86-0946-47ad-9a44-6f69c8009693) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6298 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3946 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4387 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115574 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40145 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146539 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8125 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40713 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112388 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8141 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6826 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112745 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6209 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118252 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20983 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8173 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36315 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7888 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71732 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8112 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7965 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/ImageDocument.cpp:
```diff
@@ -159,8 +160,8 @@ void ImageDocument::updateDuringParsing()
-    if (RefPtr<FragmentedSharedBuffer> buffer = loader()->mainResourceData()) {
-        if (auto* cachedImage = m_imageElement->cachedImage())
+    if (RefPtr buffer = protectedLoader()->mainResourceData()) {
+        if (auto* cachedImage = Ref { *m_imageElement }->cachedImage())
```
The static analysis doesn't warn about this but could you please use `CachedResourceHandle` instead of `auto*`?

annevk replied:
Done for all four instances I found in this file.

cdumez commented on Source/WebCore/html/ImageDocument.cpp:
```diff
@@ -169,13 +170,19 @@ void ImageDocument::updateDuringParsing()
+    }
+
+    if (RefPtr imageElement = m_imageElement.get(); imageElement && imageElement->cachedImage()) {
+        auto& cachedImage = *imageElement->cachedImage();
```
Ditto about using `CachedResourceHandle`.

cdumez reviewed (approved):
LGTM with nits.

rniwa commented on Source/WebCore/html/ImageDocument.cpp:
```diff
@@ -169,13 +170,19 @@ void ImageDocument::updateDuringParsing()
-        CachedImage& cachedImage = *m_imageElement->cachedImage();
-        RefPtr<FragmentedSharedBuffer> data = loader()->mainResourceData();
+    if (parser()->isStopped()) {
+        HTMLDocument::finishedParsing();
```
Where did this code come from??

cdumez replied:
From this end of this function, he just duplicated one line of code in order to do an early return here.

annevk committed:
Reduce unsafeness in ImageDocument
https://bugs.webkit.org/show_bug.cgi?id=304321

Reviewed by Chris Dumez and Ryosuke Niwa.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines and
adopt CachedResourceHandle while here.

Canonical link: https://commits.webkit.org/304649@main
